File: SJF algorithm/sjf.py
import operator  # operator for sorting

# creating a simple class structure for incoming data about processes
class Process:
    def __init__(self, number, arrival, duration):
        self.number = number  # process number
        self.arrival = arrival  # process arrival time
        self.duration = duration  # process cpu usage duration

    def __str__(self):
        return str(self.__class__) + ": " + str(self.__dict__)


def waiting_time(container, count):
    # our starting time will be the first arrival time
    time = container[0].arrival
    # waiting time and total time, wt for each iteration tt for total
    wt, tt = 0, 0

    for i in range(count):
        proc = container[i]  # taking a process from the process container

        if proc.arrival >= time:  # if the arrival time is after the process is done
            wt = 0
            time += proc.arrival - time
        else:
            wt = time - proc.arrival

        time += proc.duration
        tt += wt

    return tt


def sorting(container, count):  # for sorting with arrival and later with the duration

    container.sort(key=operator.attrgetter('arrival'))  # sorting for arrival first
    max_count, use_time, j = count, 0, 0

    for i in range(1, count):

        use_time += container[i - 1].duration

        max_count = count
        for j in range(i, count):
            print(j)
            if use_time < container[j].arrival:
                max_count = j
                break

        # then sorting for duration
        container[0 + i: 0 + max_count] = sorted(container[0 + i: 0 + max_count], key=operator.attrgetter('duration'))

    return container


if __name__ == '__main__':

    processes, answer = [], 0

    with open('sjf.inp') as input_fp, open('sjf.out', 'w') as output_fp:

        process_count = int(input_fp.readline())  # reads the first line which contains the process count

        for line in input_fp:
            num, arr, dur = map(int, line.split())
            processes.append(Process(num, arr, dur))  # the incoming data is divided into three parts

        answer = waiting_time(sorting(processes, process_count), process_count)
        output_fp.write(str(answer))

        print(vars(Process))

File: SJF algorithm/sjf.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

FILE* input_fp, * output_fp;

typedef vector<int> process; // a simple structure for containing process data
// 0: process number 1: arrival time 2: cpu usage time

void waitingTime(vector<process> vec, int processCount)
{
	int time = vec[0][1], wt = 0, tt = 0;//our starting time will be the first arrival time
	// waiting time and total time, wt for each iteration tt for total
	for (int i = 0; i < processCount; i++)
	{
		if (vec[i][1] >= time)// if the arrival is after all the processes are done
		{
			wt = 0;
			time += vec[i][1] - time;
		}
		else wt = time - vec[i][1];

		time += vec[i][2];
		tt += wt;
	}

	fprintf(output_fp, "%d", tt);
}

vector<process> sorting(vector<process> storeVector, int processCount) // function for sorting
{
	sort(storeVector.begin(), storeVector.end(), [](const std::vector< int >& a, const std::vector< int >& b) { return a[1] < b[1]; }); // sorting for arrival first

	int maxCount = processCount, useTime = 0;


	for (int i = 1; i < processCount; i++)
	{
		useTime += storeVector[i - 1][2];
		int j;
		for (j = i; j < processCount; j++)
		{
			if (useTime < storeVector[j][1])
			{
				maxCount = j;
				break;
			}
		}
		if (j == processCount)
			maxCount = processCount;
		sort(storeVector.begin() + i, storeVector.begin() + maxCount, [](const std::vector< int >& a, const std::vector< int >& b) { return a[2] < b[2]; }); // sorting for usage
	}
	return storeVector;
}

int main()
{
	if ((input_fp = fopen("sjf.inp", "rt")) == NULL)
	{
		fprintf(stderr, "can't open %s\n", "sjf.inp");
	}
	if ((output_fp = fopen("sjf.out", "wt")) == NULL)
	{
		fprintf(stderr, "can't open %s\n", "sjf.out");
	}

	int processCount;//number of processes to be added

	vector<process> storeVector;//vector containing all the processes

	fscanf(input_fp, "%d", &processCount);


	for (int i = 0; i < processCount; i++)
	{
		process process;
		for (int j = 0; j < 3; j++)
		{
			int temp = 0;
			fscanf(input_fp, "%d", &temp);
			process.push_back(temp);// process number, arrival and usage time are added
		}
		storeVector.push_back(process);// the process is then appended to storeVector
	}
	
	waitingTime(sorting(storeVector, processCount) , processCount);// sorts and computes waiting time
}